vpr/src/route/rr_graph_fpga_interchange.cpp

```diff
@@ -0,0 +1,1706 @@
+    }
+
+    /*
+     * Debug print function
+     */
+    void print_virtual() {
+        VTR_LOG("Virtual\n");
+        for (const auto& entry : sink_source_loc_map_) {
+            const auto& key = entry.first;
+            const auto& value = entry.second;
+            int x, y;
+            std::tie(x, y) = tile_to_loc_[key];
+            VTR_LOG("Location x:%d y:%d Name:%s\n", x, y, str(key).c_str());
+            int it = 0;
+            for (const auto& pin : value) {
+                bool dir;
+                float RC;
+                int node_id;
+                std::tie(dir, RC, node_id) = pin;
+                VTR_LOG("\tpin:%d input:%d R/C:%e\n", it, dir, RC);
+                VTR_LOG("\ttile_name:%s node_id:%d\n", str(key).c_str(), node_id);
+                it++;
+            }
+        }
+
+        int it = 0;
+
+        for (auto const& switch_id : device_ctx_.rr_switch_inf) {
+            VTR_LOG("Switch: %d Name:%s\n", it++, switch_id.name);
+        }
+
+        for (auto& entry : virtual_chan_loc_map_) {
+            location loc, loc2;
+            e_rr_type type;
+            std::tie(loc, type) = entry.first;
+            VTR_LOG("CHAN%c X:%d Y:%d\n", type == e_rr_type::CHANX ? 'X' : 'Y', loc.first, loc.second);
+            for (auto& seg : entry.second) {
+                loc2 = std::get<3>(seg);
+                VTR_LOG("\tSegment id:%d name:%s -> X:%d Y:%d\n", std::get<0>(seg), segment_inf_[RRSegmentId(std::get<0>(seg))].name.c_str(), loc2.first, loc2.second);
+            }
+        }
+
+        VTR_LOG("Redirects:\n");
+        for (auto& entry : virtual_redirect_) {
+            int node;
+            location loc;
+            std::tie(node, loc) = entry.first;
+
+            location new_loc;
+            e_rr_type new_type;
+            int new_idx;
+            std::tie(new_loc, new_type, new_idx) = entry.second;
+            VTR_LOG("\t Node:%d X:%d Y:%d -> X:%d Y:%d CHAN%c idx:%d\n", node, loc.first, loc.second, new_loc.first,
+                    new_loc.second, new_type == e_rr_type::CHANX ? 'X' : 'Y', new_idx);
+        }
+
+        VTR_LOG("Shorts:\n");
+        for (auto& entry : virtual_shorts_) {
+            location loc1, loc2;
+            e_rr_type type1, type2;
+            int id1, id2;
+            std::tie(loc1, type1, id1, loc2, type2, id2) = entry;
+            VTR_LOG("\tCHAN%c X:%d Y%d", type1 == e_rr_type::CHANX ? 'X' : 'Y', loc1.first, loc1.second);
+            VTR_LOG(" Segment id:%d name:%s ->", id1,
+                    segment_inf_[RRSegmentId(std::get<0>(virtual_chan_loc_map_[std::make_tuple(loc1, type1)][id1]))].name.c_str());
+            VTR_LOG(" CHAN%c X:%d Y:%d", type2 == e_rr_type::CHANX ? 'X' : 'Y', loc2.first, loc2.second);
+            VTR_LOG(" Segment id:%d name:%s\n", id2,
+                    segment_inf_[RRSegmentId(std::get<0>(virtual_chan_loc_map_[std::make_tuple(loc2, type2)][id2]))].name.c_str());
+        }
+    }
+
+    void print() {
```

Comment:
It is not entirely clear what is the difference between the real and the virtual graph. maybe some comments around it can help clarify it

Reply:
Explanation added in next commit.